apollo-router/src/plugins/healthcheck/mod.rs

```diff
@@ -0,0 +1,506 @@
+        let my_ready = ready.clone();
+
+        my_rejected.store(0, Ordering::Relaxed);
+        let ticker = tokio::spawn(async move {
+            loop {
+                let start = Instant::now() + my_sampling_interval;
+                let mut interval = tokio::time::interval_at(start, my_sampling_interval);
+                loop {
+                    interval.tick().await;
+                    if my_rejected.load(Ordering::Relaxed) > allowed {
+                        my_ready.store(false, Ordering::SeqCst);
+                        tokio::time::sleep(my_recovery_interval).await;
+                        my_rejected.store(0, Ordering::Relaxed);
+                        my_ready.store(true, Ordering::SeqCst);
+                        break;
+                    }
+                }
+            }
+        });
+        Ok(Self {
+            config: init.config,
+            live,
+            ready,
+            rejected,
+            ticker,
+        })
+    }
+
+    // Track rejected requests due to traffic shaping.
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        let my_rejected = self.rejected.clone();
+
+        ServiceBuilder::new()
+            .map_response(move |res: router::Response| {
+                if res.response.status() == StatusCode::SERVICE_UNAVAILABLE
+                    || res.response.status() == StatusCode::GATEWAY_TIMEOUT
+                {
+                    my_rejected.fetch_add(1, Ordering::Relaxed);
+                }
+                res
+            })
+            .service(service)
+            .boxed()
+    }
+
+    // Support the health-check endpoint for the router, incorporating both live and ready.
+    fn web_endpoints(&self) -> MultiMap<ListenAddr, Endpoint> {
+        let mut map = MultiMap::new();
+
+        let my_ready = self.ready.clone();
+        let my_live = self.live.clone();
+
+        let endpoint = Endpoint::from_router_service(
+            self.config.path.clone(),
+            service_fn(move |req: router::Request| {
+                let mut status_code = StatusCode::OK;
+                let health = if let Some(query) = req.router_request.uri().query() {
+                    let query_upper = query.to_ascii_uppercase();
+                    // Could be more precise, but sloppy match is fine for this use case
+                    if query_upper.starts_with("READY") {
+                        let status = if my_ready.load(Ordering::SeqCst) {
+                            HealthStatus::Up
+                        } else {
+                            // It's hard to get k8s to parse payloads. Especially since we
+                            // can't install curl or jq into our docker images because of CVEs.
+                            // So, compromise, k8s will interpret this as probe fail.
+                            status_code = StatusCode::SERVICE_UNAVAILABLE;
+                            HealthStatus::Down
+                        };
```

Comment:
I can't match up this comment with this if-else. So if `my_ready` is at [AtomicBool] of 1, so health status is up, if at 0, we are down. But what does the payload have to do with the health status being down? is it that for a specific type of payload we declare somewhere else that the health check is not ready?

Reply:
The payload is the textual representation of the health status, the status code is the HTTP status code. k8s will drive its probe behaviour from the HTTP status code, but other clients may choose to parse the json body that we return. We are basically returning two different indicators of up/down-ness: status code OK or UNAVAILABLE and some body text (JSON) which is UP or DOWN (matching the status codes). I'm not sure this is answering your question...